src/Symfony/Component/Debug/DebugClassLoader.php

```diff
@@ -522,4 +557,149 @@ private function getOwnInterfaces($class, $parent)
+            return;
+        }
+
+        if (!preg_match_all('/\n\s+\* @return +(\S+)/', $doc, $matches, PREG_SET_ORDER)) {
+            return;
+        }
+
+        $types = explode('|', end($matches)[1], 3);
+
+        $nullable = false;
+
+        switch (\count($types)) {
+            case 1:
+                $type = $types[0];
+
+                break;
+            case 2:
+                // fastest way to handle the null|null edge case
+                $type = 'null';
+
+                foreach (array_unique(array_map([$this, 'normalizeType'], $types)) as $i => $normalizedType) {
+                    if ('null' === $normalizedType) {
+                        $nullable = true;
+                    } else {
+                        $type = $types[$i];
+                    }
+                }
+
+                if (!$nullable) {
+                    return;
+                }
+
+                break;
+            default:
+                return;
+        }
+
+        $normalizedType = $this->normalizeType($type);
+        if (\in_array($normalizedType, self::COMMON_NON_OBJECT_RETURNED_TYPES)) {
+            return;
+        }
+
+        if (isset(self::NULLABLE_RETURNABLE_TYPES[$normalizedType])) {
+            $normalizedDocReturnType = self::NULLABLE_RETURNABLE_TYPES[$normalizedType];
+            $hintedReturnType = $normalizedType;
+        } elseif (isset(self::NON_NULLABLE_RETURNABLE_TYPES[$normalizedType])) {
+            if ($nullable) {
+                return;
+            }
+
+            $normalizedDocReturnType = self::NON_NULLABLE_RETURNABLE_TYPES[$normalizedType];
+            $hintedReturnType = $normalizedType;
+        } else {
+            $normalizedDocReturnType = sprintf('an instance of "%s"', $type);
+            $hintedReturnType = $type;
+        }
+
+        if ($nullable) {
+            $normalizedDocReturnType .= ' or null';
+            $hintedReturnType = '?'.$hintedReturnType;
+        }
+
+        self::$returnTypes[$class][$method->name] = [$normalizedDocReturnType, $hintedReturnType];
+    }
+
+    /**
+     * @param string           $class
+     * @param string|false     $parent
+     * @param \ReflectionClass $refl
+     * @param array            $deprecations
+     */
+    private function checkReturnTypes(string $class, $parent, \ReflectionClass $refl, &$deprecations): void
+    {
+        $parentsAndInterfacesThatHaveMethodsThatReturnsADocType = [];
+        foreach (array_merge($parent ? array_merge([$parent], class_parents($parent, false)) : [], class_implements($class, false)) as $use) {
+            if (isset(self::$returnTypes[$use])) {
+                $parentsAndInterfacesThatHaveMethodsThatReturnsADocType[] = $use;
+            }
+        }
+
+        if (empty($parentsAndInterfacesThatHaveMethodsThatReturnsADocType)) {
+            return;
+        }
+
+        foreach ($refl->getMethods(\ReflectionMethod::IS_PUBLIC | \ReflectionMethod::IS_PROTECTED) as $method) {
```

Comment:
this means we'll create twice the list of methods, but this is a perf-critical class, let's create those reflectionmethod instances once